pkg/controllers/termination/terminator/eviction.go

```diff
@@ -56,61 +60,79 @@ func IsNodeDrainError(err error) bool {
 }
 
-func NewEvictionQueue(ctx context.Context, coreV1Client corev1.CoreV1Interface, recorder events.Recorder) *EvictionQueue {
-	queue := &EvictionQueue{
+func NewQueue(coreV1Client corev1.CoreV1Interface, recorder events.Recorder) *Queue {
+	queue := &Queue{
 		RateLimitingInterface: workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(evictionQueueBaseDelay, evictionQueueMaxDelay)),
 		Set:                   set.NewSet(),
 		coreV1Client:          coreV1Client,
 		recorder:              recorder,
 	}
-	go queue.Start(logging.WithLogger(ctx, logging.FromContext(ctx).Named("eviction")))
 	return queue
 }
 
-// Add adds pods to the EvictionQueue
-func (e *EvictionQueue) Add(pods ...*v1.Pod) {
+func (q *Queue) Name() string {
+	return "eviction-queue"
+}
+
+func (q *Queue) Builder(_ context.Context, m manager.Manager) controller.Builder {
+	return controller.NewSingletonManagedBy(m)
+}
+
+// Add adds pods to the Queue
+func (q *Queue) Add(pods ...*v1.Pod) {
 	for _, pod := range pods {
-		if nn := client.ObjectKeyFromObject(pod); !e.Set.Contains(nn) {
-			e.Set.Add(nn)
-			e.RateLimitingInterface.Add(nn)
+		if nn := client.ObjectKeyFromObject(pod); !q.Set.Contains(nn) {
+			q.Set.Add(nn)
+			q.RateLimitingInterface.Add(nn)
 		}
 	}
 }
 
-func (e *EvictionQueue) Start(ctx context.Context) {
-	for {
-		// Get pod from queue. This waits until queue is non-empty.
-		item, shutdown := e.RateLimitingInterface.Get()
-		if shutdown {
-			break
-		}
-		nn := item.(types.NamespacedName)
-		// Evict pod
-		if e.evict(ctx, nn) {
-			e.RateLimitingInterface.Forget(nn)
-			e.Set.Remove(nn)
-			e.RateLimitingInterface.Done(nn)
-			continue
-		}
-		e.RateLimitingInterface.Done(nn)
-		// Requeue pod if eviction failed
-		e.RateLimitingInterface.AddRateLimited(nn)
+// Pop returns a namespaced name from the queue and if the queue has shutdown.
+func (q *Queue) Pop() (types.NamespacedName, bool) {
+	// Get pod from queue. This waits until queue is non-empty.
+	item, shutdown := q.RateLimitingInterface.Get()
+	if shutdown {
+		return types.NamespacedName{}, true
 	}
-	logging.FromContext(ctx).Errorf("EvictionQueue is broken and has shutdown")
+	nn := item.(types.NamespacedName)
+	return nn, false
 }
 
-// evict returns true if successful eviction call, and false if not an eviction-related error
-func (e *EvictionQueue) evict(ctx context.Context, nn types.NamespacedName) bool {
+func (q *Queue) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	nn, shutdown := q.Pop()
+	if shutdown {
+		return reconcile.Result{}, fmt.Errorf("queue has shutdown")
+	}
+	q.ProcessItem(ctx, nn)
+	return reconcile.Result{RequeueAfter: controller.Immediately}, nil
+}
+
+func (q *Queue) ProcessItem(ctx context.Context, nn types.NamespacedName) {
+	// Evict pod
+	if q.Evict(ctx, nn) {
+		q.RateLimitingInterface.Forget(nn)
+		q.Set.Remove(nn)
+		q.RateLimitingInterface.Done(nn)
+		return
+	}
+	q.RateLimitingInterface.Done(nn)
```

Comment:
I'm forgetting how the rate limiting interface handles this, but you call Done() to finish pulling an element out of the queue and then you add it back to the back of the queue? Is that correct? Does it make sense to call Done() after the initial Get() if it's a Pop() operation?

Reply:
It's used as a way to mark the item as done processing. When you do Get() for the item, it pulls it out of the queue and removes it from the dirty (needing processing) set. Using done will make sure if the item gets re-enqueued while we're processing that it doesn't get requeued earlier than it should, putting it at the back of the queue. 


// Done marks item as done processing, and if it has been marked as dirty again
// while it was being processed, it will be re-added to the queue for
// re-processing.

Reply:
Ah, got it. Then looks correct. I see that controller-runtime uses a defer statement for the `Done()` call. Is that something that we should consider doing?

Reply:
Yeah, defer Done() sounds right will do.